_posts/2019-09-13-applying-code-splitting-aws-lambda.md

```diff
@@ -0,0 +1,180 @@
+
+When I started thinking about this, I was looking at ways to write a serverless application that could delete resources that had failed to be cleaned up after running integration tests. In that scenario, I would need to import multiple different clients from the AWS SDK to handle different resource types - that said, I don't know which clients need to be imported until the handler executes. Let's step through this example.
+
+Imagine a project structure like the following;
+
+```
+├── DeleteResources
+│   ├── index.ts
+│   ├── lib.ts
+│   ├── resourcegroups
+│   │   └── group.ts
+│   ├── s3
+│   │   └── bucket.ts
+│   └── servicecatalog
+│       └── cloudformationproduct.ts
+```
+
+In this case, the `index.ts` file contains the handler logic. It consumes entries from an SQS queue that contain the resource type and the resource ARN. The folder structure for the resources is deliberately designed to match the resource type records to make it easy to work out what file needs to be imported - e.g. `AWS::S3::Bucket` would map to the file in `s3/bucket`.
+
+```typescript
+import { SQSHandler, SQSRecord } from "aws-lambda";
+import { Delete } from "./lib";
+
+interface Payload {
+    ResourceArn: string
+    ResourceType: string
+}
+
+async function DeleteRecord (record: SQSRecord): Promise<void> {
+    const { ResourceType, ResourceArn } = JSON.parse(record.body) as Payload;
+    return await Delete(ResourceType, ResourceArn);
+}
+
+export const handler: SQSHandler = async (event) => {
+    await Promise.all(event.Records.map(DeleteRecord));
+}
+```
+
+The delete function in `./lib` is where the magic happens. As is turns out, you can take advantage of code-splitting functionality in webpack to do the heavy-lifting. It looks like so;
+
+```typescript
+export const Delete = async (type: string, arn: string): Promise<void> => {
+    const [ _, service, resource ] = type.split("::").map(e => e.toLocaleLowerCase());
+    const mod = await import(
+        /* webpackInclude: /(s3|servicecatalog|resourcegroups)/ */
+        /* webpackMode: "lazy-once" */
+        `./${service}/${resource}`);
+
+    await mod.default(arn);
+}
+```
+
+In this case, I use a dynamic import statement with additional metadata to tell webpack how to handle the dynamic import. The `webpackInclude` statement tells webpack that I will only ever try to load files that match the following regex at runtime. The argument to the import statement already tells webpack that it needs to look for possible candidate files in the current directory. Finally, the `webpackMode` directive tells webpack that I want to lazy-load the module once on usage, and keep it cached in case the module needs to be resolved again.
```

Comment:
Why is the import statement important?

Link to Serverless Security Top 10?

Reply:
It doesn't need to be called out explicitly - just a link for those that want to do further reading.

Reply:
> Why is the import statement important?
> 
> Link to Serverless Security Top 10?

It's not immediately obvious to me where you are going with this?